ibmcloud/SECURE_EXECUTION.md

```diff
@@ -0,0 +1,102 @@
+
+## Set up a demo environment without Secure Execution
+
+Follow the [README.md](./README.md) to setup a demo environment on IBM Cloud VPC. We need create a **s390x** architecture based cluster, so please include the following two lines in the `terraform.tfvars` file:
+```
+instance_profile_name = "bz2-2x8"
+image_name = "ibm-ubuntu-18-04-1-minimal-s390x-3"
+```
+
+## Prepare worker node to build SE enabled image
+
+1. **Build genprotimg**
+
+`genprotimg` is used to build an encrypted boot record from a given kernel, initial RAM disk, parameters, and public host-key document.
+
+For Ubuntu 18.04, we can build it from the source on GitHub:
+```bash
+git clone https://github.com/ibm-s390-linux/s390-tools.git
+cd s390-tools/genprotimg/
+apt install -y libglib2.0-dev libssl-dev libcurl4-openssl-dev
+make install
+```
+
+2. **Download host key document from Resource Link**
+
+The host key must match the host system for which the image is prepared. Please refer to document [Obtaining a host key document from Resource Link](https://www.ibm.com/docs/en/linux-on-systems?topic=execution-obtain-host-key-document#lxse_obtain_hkd) for details.
+
+As a registered user, access the search page:
+```
+https://www.ibm.com/servers/resourcelink/hom03010.nsf/pages/HKDSearch?OpenDocument
+```
+
+Need input the **machine type** and **machine serial number**, which can be obtained from `/proc/sysinfo` of the s390x worker node. Check the values of **Type** and **Sequence Code**. Machine serial number would be the last 5 or 7 characters. Please refer to [this documentation](https://www.ibm.com/docs/en/linux-on-systems?topic=tasks-find-machine-serial) for details.
+
+When you obtain the host key document, please send it to the worker node host keys documents directory.
+
+> **NOTE**: You can download multiple different host keys and prepare only one SE enabled image for different host systems.
+
+## Build the SE enabled qcow2 image
+
+Still use the [build script](./image/build.sh) for IBM cloud image. To build SE enabled image, please set `SE_BOOT` as 1, and set `HOST_KEYS_DIR` with the host key documents directory. For example:
+```
+cd ibmcloud/image
+CLOUD_PROVIDER=ibmcloud SE_BOOT=1 HOST_KEYS_DIR=/root/ make build
+```
+> **Note** The [build script](./image/build.sh) is following the steps in [the document](https://www.ibm.com/docs/en/linux-on-systems?topic=execution-workload-owner-tasks)
+
+## Push the SE enabled qcow2 image to IBM Cloud COS bucket
+
+Use the [push script](./image/push.sh) to push the SE enabled qcow2 image to IBM Cloud COS bucket. Please confirm the IBM Cloud API key, COS instance name and COS bucket name when you set up the demo environment:
+```
+export IBMCLOUD_API_KEY=<your_api_key>
+export IBMCLOUD_COS_SERVICE_INSTANCE=<your_cos_service_instance_name>
+export IBMCLOUD_COS_BUCKET=<your_cos_bucket_name>
+./push.sh --name se-podvm-566065e-dirty-s390x.qcow2 --path se-podvm-566065e-dirty-s390x.qcow2
```

Comment:
The exported variables are only going to be valid if the user is running the demo environment inside `jp-tok-1` As this is the default for the script. 
I think adding a note to mention that exported variables will need to expanded on to include: 
```
export CLOUD_PROVIDER=ibmcloud
export IMAGE_NAME=<An image name with arch included e.g. image-amd64>
export IBMCLOUD_COS_REGION=<the region your IKS cluster and zVSI is running on>
export IBMCLOUD_VPC_REGION=$IBMCLOUD_COS_REGION
export IBMCLOUD_VPC_NAME=<your generated vpc name from the terraform>
export IBMCLOUD_VPC_SUBNET_NAME=<the linked subnet name to the above vpc>
export IBMCLOUD_COS_SERVICE_ENDPOINT="https://s3.${IBMCLOUD_COS_REGION}.cloud-object-storage.appdomain.cloud"
export IBMCLOUD_COS_SERVICE_INSTANCE=<your cos instance>
export IBMCLOUD_COS_BUCKET=<your cos bucket in the region your cluster runs in>
export IBMCLOUD_API_KEY=<your ibmcloud apikey>
export IBMCLOUD_API_ENDPOINT="https://cloud.ibm.com"
export GOPATH=/root/go
``` 
And I think that `./push.sh` command needs to remove the `.qcow2` from the name as it seems to fail due to it being an invalid name.

Reply:
OK, push.sh will directly create custom image after uploading the qcow2 file to COS bucket. Here I just need upload the file to bucket. I will change the document to use `multipart_upload.sh` instead.

Reply:
OK, so the multipart upload will only create the image inside the COS bucket. And we want the user to manually create the custom image in ibmcloud. 

The main difference I can see is the change of OS, to point to hyper-protect instead when creating the custom image. 
Could we instead look at adding something similar to: 
```
[ "$SE_BOOT" = "1" ] && os_name="hyper-protect-1-0-s390x" || os_name="ubuntu-20-04-${arch/x86_64/amd64}"
```
To this line: [ibmcloud/image/push.sh#L96](https://github.com/liudalibj/cloud-api-adaptor/blob/6af27a7789a8a7624daf58fccb90f309fda7281e/ibmcloud/image/push.sh#L96)

And remove the need to separate the different `build` then `upload` then `create` steps? And just run: `CLOUD_PROVIDER=ibmcloud SE_BOOT=1 HOST_KEYS_DIR=/root/ make push`. This would mean that the requested changes mentioned here: https://github.com/confidential-containers/cloud-api-adaptor/pull/407#discussion_r1046085254. Will also be needed.

Reply:
I will updated it.